api/src/main/java/org/semanticweb/owlapi/model/axiomproviders/EquivalentAxiomProvider.java

```diff
@@ -41,6 +41,17 @@
         return getOWLEquivalentClassesAxiom(classExpressions, Collections.emptySet());
     }
 
+    /**
+     * @param classExpressions equivalent classes. Cannot be null or contain nulls.
+     * @param keepDuplicates whether to keep duplicate items
+     * @return an equivalent classes axiom with specified operands and no annotations
+     */
+    default OWLEquivalentClassesAxiom
+    getOWLEquivalentClassesAxiom(Collection<? extends OWLClassExpression> classExpressions,
+                                 boolean keepDuplicates) {
```

Comment:
This is one possible solution to the reasoner need for tautologies, but it requires the reasoner to be changed to call these methods. Ideally, a solution that can be switched on or off outside the reasoner code would be better. Given that data factories do not come with their own configuration, we're probably restricted to system properties (along the lines of `ConfigurationOptions` anyway). I'm not sure which solution is best at this point.

I'm also not totally convinced only two axiom types need this ability; although those are the ones mentioned so far, removing duplicates is currently done to more axiom types than that (subclass, equivalent classes and properties, disjoint classes and properties, disjoint union, off the top of my head). I think it would be easier to have an option to switch off duplicate removal for all axiom types, not just one.

Reply:
> (subclass, equivalent classes and properties, disjoint classes and properties, disjoint union, off the top of my head). I think it would be easier to have an option to switch off duplicate removal for all axiom types, not just one.

Yes, that'd be easier! My current understanding is that all these classes ultimately call the sorted function in OWLAPIStreamUtils.java, which does the duplicate removal. A simple solution can be to set a static parameter "keepDuplicates" set on or off in this util class, which will switch on or off the duplicate removal done by this class. In addition to this, the interfaces would need to give up creating sets of the input arguments and use lists instead. 

This would solve both problems - ability to switch on/off the duplicate removal from outside the reasoner and allowing duplicate removal for the classes which use StreamUtils to create the sorted list of input arguments (which seem to be most of them). 

I originally was trying the above but was under the impression we'd need to allow duplicated only to those axioms which actually need duplicates. Rather than disabling duplicate removal for _all_ the classes, as they might actually require removing them. So, I added the duplicate removal as an additional (optional) call in the StreamUtils, and limited its use to specific interface/implementation.


So should I attempt the above mentioned solution?

Reply:
> Yes, that'd be easier! My current understanding is that all these classes ultimately call the sorted function in OWLAPIStreamUtils.java, which does the duplicate removal. A simple solution can be to set a static parameter "keepDuplicates" set on or off in this util class, which will switch on or off the duplicate removal done by this class. In addition to this, the interfaces would need to give up creating sets of the input arguments and use lists instead.

Hello everyone.

The solution with "OWLAPIStreamUtils.java + static keepDuplicate flag" has limited implementation capabilities - it can be implemented only in the OWL-API-impl.
Different implementations, such as ONT-API, don't even use `OWLAPIStreamUtils` and static flags.  
In my opinion, the better solution is to add config option to the `OWLOntologyLoaderConfiguration`&`OntologyConfigurator`. 
These two classes are parts of OWL-API-api (with is a contract that describes behaviour), altough they are concret classes, not interfaces, unfortunately. 
But that means a lot more work

Reply:
https://github.com/owlcs/owlapi/commit/b51e6930bfb26f5bfabe0701c2408782385cd894 this should do the job, it allows the behaviour to be defined fr the whole VM and all factories created within it (with no changes to client code) or to create one factory with different behaviour (no impact on other functionalities but it requires changes to the reasoner).

Reply:
Update: that's the first commit for version5; for version4, that's all that can be done without changing the API (i.e., OWLDataFactory uses only Sets as inputs, in version4).